enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,368 @@
+   including init and ephemeral containers.
+
+Per upstream docs, the mutating admission webhook will need to register itself with the cluster and
+run as a service in its own namespace [1]. The operator installing the Subscription Injection
+webhook will need to do the following:
+
+1. Create a `Deployment` that runs the webhook in a highly available mode.
+2. Create a `Service` to expose the webhook pods, with generated TLS cert/key pairs.
+3. Create a `ConfigMap` where the service serving CA can be injected.
+4. Create a `MutatingWebhookConfiguration` that fires the webhook for all created pods.
+5. When the service serving CA in item 3 is injected or updated, the operator needs to update the
+   caBundle provided to the webook configuration in 4.
+
+[1]
+https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#mutating-webhook-configuration
+
+#### Bundle API
+
+```yaml
+kind: ClusterBundle
+apiVersion: subscription.openshift.io/v1alpha1
+metadata:
+  name: <name>
+  ... # standard metadata, cluster-scoped so no namespace
+spec:
+  entitlement:
+    name: <secret name with entitlement keys, must be in openshift-config namespace>
+  yumRepository:
+    name: <configMap name with yum repo definition, must be in openshift-config namespace>
+  aggregateToClusterRoles: # cluster roles which the ClusterBundle RBAC should aggregate to.
+                           # This can override the operator default
+  - edit
+  - admin
+status:
+  conditions:
+  - type: Invalid
+    status: "False" # standard k8s conditions, should obey the "abnormal-true" convention
+  - ... # additional informational conditions
+  clusterRole: <name>-subscription-bundle # clusterRole is cluster-scoped
+  shares:
+  - name: <name>-subscription-entitlement # shares are cluster-scoped
+  - name: <name>-subscription-yumRepository
+```
+
+#### Operator API
+
+```yaml
+kind: Operator
+apiVersion: subscription.openshift.io/v1alpha1
+metadata:
+  name: cluster # canonical
+  namespace: openshift-subscription-injection
+spec:
+  defaultClusterBundles: # default cluster bundles installed by the operator
+    - name: rhel7
+      namespace: openshift-config
+      entitlement: etc-pki-entitlement
+      yumRepository: redhat-repo-rhel7
+    - name: rhel8
+      namespace: openshift-config
+      entitlement: etc-pki-entitlement
+      yumRepository: redhat-repo-rhel8
+  defaultAggregatedClusterRoles: # default cluster roles that the ClusterBundle RBAC aggregates to
+    - edit
+    - admin
+status:
+  conditions: # standard status conditions, can be added as needed.
+    - type: Available
+      status: "True"
+```
+
+### Risks and Mitigations
+
+**Risk:** Entitlement keys can leak outside of the `openshift-config` namespace to unauthorized
+workloads.
+
+_Mitigation:_ Access to the entitlement keys is gated by the cluster role generated for the
+projected resource shares associated with the bundle. Cluster admins are responsible for either
```

Comment:
not clear how you're gating access, since someone can directly go read the entitlement configmap in the openshift-config namespace.  I thought in general we said things in that namespace should not be considered private.

Perhaps this content needs to live elsewhere, unless you can show we are already putting private values into openshift-config.

Reply:
Currently the cluster pull secret lives as secret in `openshift-config` - in customer clusters this includes the `cloud.redhat.com` auth token that subscribes the cluster and enables telemetry. Deleting this token is how customers opt-out of telemetry [1].

The entiltement key would live as an opaque Secret (not a ConfigMap) - is this any more sensitive than the pull secret auth tokens?

[1] https://docs.openshift.com/container-platform/4.5/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html#insights-operator-new-pull-secret_opting-out-remote-health-reporting

Reply:
ok, i think it's fine then